src/platforms/stm32/src/lib/gpio_driver.c

```diff
@@ -229,6 +229,15 @@ static term create_pair(Context *ctx, term term1, term term2)
     return ret;
 }
 
+static term create_error_tuple(Context *ctx, term err_atom)
+{
+    if (UNLIKELY(memory_ensure_free_with_roots(ctx, 3, 1, &err_atom, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
```

Comment:
If `err_atom` is always an atom, and if the call to `create_error_tuple` is always returning and not resulting in a new allocation (I don't know if this is true, so you might need to check), then the `_with_roots` is unnecessary, since ``err_atom` will never be GC'd.

You might want to add a check to ensure that `err-atom` is always an atom, or at least document so, so that future maintainers don't pass in a tuple expecting things to work.

Reply:
Thank you for the clarification, I thought there was a possibility of `err_atom` being GC'd if `with_roots` was not used. And I should absolutely add a check that `err_atom` is actually an atom.

Reply:
Fixed.